Doesn't build on node 0.11.13 (MacOS)

npm i gaze

```
> gaze@0.6.4 install /js/javascript-nodejs/node_modules/gaze
> node build.js

  CXX(target) Debug/obj.target/pathwatcher/src/main.o
In file included from ../src/main.cc:21:
In file included from ../src/common.h:26:
../node_modules/nan/nan.h:339:13: error: no member named 'New' in 'v8::String'
    return  _NAN_ERROR(v8::Exception::Error, errmsg);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:319:50: note: expanded from macro '_NAN_ERROR'
# define _NAN_ERROR(fun, errmsg) fun(v8::String::New(errmsg))
                                     ~~~~~~~~~~~~^
../node_modules/nan/nan.h:343:5: error: no member named 'ThrowException' in namespace 'v8'
    _NAN_THROW_ERROR(v8::Exception::Error, errmsg);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:324:11: note: expanded from macro '_NAN_THROW_ERROR'
      v8::ThrowException(_NAN_ERROR(fun, errmsg));                             \
      ~~~~^
../node_modules/nan/nan.h:343:5: error: no member named 'New' in 'v8::String'
    _NAN_THROW_ERROR(v8::Exception::Error, errmsg);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:324:26: note: expanded from macro '_NAN_THROW_ERROR'
      v8::ThrowException(_NAN_ERROR(fun, errmsg));                             \
                         ^~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:319:50: note: expanded from macro '_NAN_ERROR'
# define _NAN_ERROR(fun, errmsg) fun(v8::String::New(errmsg))
                                     ~~~~~~~~~~~~^
../node_modules/nan/nan.h:348:9: error: no type named 'ThrowException' in namespace 'v8'
    v8::ThrowException(error);
    ~~~~^
../node_modules/nan/nan.h:355:65: error: no member named 'New' in 'v8::String'
    v8::Local<v8::Value> err = v8::Exception::Error(v8::String::New(msg));
                                                    ~~~~~~~~~~~~^
../node_modules/nan/nan.h:357:26: error: no member named 'New' in 'v8::String'
    obj->Set(v8::String::New("code"), v8::Int32::New(errorNumber));
             ~~~~~~~~~~~~^
../node_modules/nan/nan.h:369:12: error: no member named 'New' in 'v8::String'
    return _NAN_ERROR(v8::Exception::TypeError, errmsg);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:319:50: note: expanded from macro '_NAN_ERROR'
# define _NAN_ERROR(fun, errmsg) fun(v8::String::New(errmsg))
                                     ~~~~~~~~~~~~^
../node_modules/nan/nan.h:373:5: error: no member named 'ThrowException' in namespace 'v8'
    _NAN_THROW_ERROR(v8::Exception::TypeError, errmsg);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:324:11: note: expanded from macro '_NAN_THROW_ERROR'
      v8::ThrowException(_NAN_ERROR(fun, errmsg));                             \
      ~~~~^
../node_modules/nan/nan.h:373:5: error: no member named 'New' in 'v8::String'
    _NAN_THROW_ERROR(v8::Exception::TypeError, errmsg);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:324:26: note: expanded from macro '_NAN_THROW_ERROR'
      v8::ThrowException(_NAN_ERROR(fun, errmsg));                             \
                         ^~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:319:50: note: expanded from macro '_NAN_ERROR'
# define _NAN_ERROR(fun, errmsg) fun(v8::String::New(errmsg))
                                     ~~~~~~~~~~~~^
../node_modules/nan/nan.h:377:12: error: no member named 'New' in 'v8::String'
    return _NAN_ERROR(v8::Exception::RangeError, errmsg);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:319:50: note: expanded from macro '_NAN_ERROR'
# define _NAN_ERROR(fun, errmsg) fun(v8::String::New(errmsg))
                                     ~~~~~~~~~~~~^
../node_modules/nan/nan.h:381:5: error: no member named 'ThrowException' in namespace 'v8'
    _NAN_THROW_ERROR(v8::Exception::RangeError, errmsg);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:324:11: note: expanded from macro '_NAN_THROW_ERROR'
      v8::ThrowException(_NAN_ERROR(fun, errmsg));                             \
      ~~~~^
../node_modules/nan/nan.h:381:5: error: no member named 'New' in 'v8::String'
    _NAN_THROW_ERROR(v8::Exception::RangeError, errmsg);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:324:26: note: expanded from macro '_NAN_THROW_ERROR'
      v8::ThrowException(_NAN_ERROR(fun, errmsg));                             \
                         ^~~~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:319:50: note: expanded from macro '_NAN_ERROR'
# define _NAN_ERROR(fun, errmsg) fun(v8::String::New(errmsg))
                                     ~~~~~~~~~~~~^
../node_modules/nan/nan.h:727:49: error: too few arguments to function call, single argument 'isolate' was not specified
    v8::Local<v8::Object> obj = v8::Object::New();
                                ~~~~~~~~~~~~~~~ ^
/Users/iliakan/.node-gyp/0.11.13/deps/v8/include/v8.h:2457:3: note: 'New' declared here
  static Local<Object> New(Isolate* isolate);
  ^
In file included from ../src/main.cc:21:
In file included from ../src/common.h:26:
../node_modules/nan/nan.h:733:49: error: too few arguments to function call, single argument 'isolate' was not specified
    v8::Local<v8::Object> obj = v8::Object::New();
                                ~~~~~~~~~~~~~~~ ^
/Users/iliakan/.node-gyp/0.11.13/deps/v8/include/v8.h:2457:3: note: 'New' declared here
  static Local<Object> New(Isolate* isolate);
  ^
In file included from ../src/main.cc:21:
In file included from ../src/common.h:26:
../node_modules/nan/nan.h:740:12: error: no member named 'Dispose' in 'v8::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> >'
    handle.Dispose();
    ~~~~~~ ^
../node_modules/nan/nan.h:741:12: error: no member named 'Clear' in 'v8::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> >'
    handle.Clear();
    ~~~~~~ ^
../node_modules/nan/nan.h:746:39: error: no member named 'NewSymbol' in 'v8::String'; did you mean 'IsSymbol'?
    NanPersistentToLocal(handle)->Set(NanSymbol("callback"), fn);
                                      ^~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:181:38: note: expanded from macro 'NanSymbol'
#define NanSymbol(value) v8::String::NewSymbol(value)
                         ~~~~~~~~~~~~^
/Users/iliakan/.node-gyp/0.11.13/deps/v8/include/v8.h:1441:8: note: 'IsSymbol' declared here
  bool IsSymbol() const;
       ^
In file included from ../src/main.cc:21:
In file included from ../src/common.h:26:
../node_modules/nan/nan.h:746:39: error: call to non-static member function without an object argument
    NanPersistentToLocal(handle)->Set(NanSymbol("callback"), fn);
                                      ^~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:181:38: note: expanded from macro 'NanSymbol'
#define NanSymbol(value) v8::String::NewSymbol(value)
                         ~~~~~~~~~~~~^~~~~~~~~
../node_modules/nan/nan.h:750:46: error: no member named 'NewSymbol' in 'v8::String'; did you mean 'IsSymbol'?
    return NanPersistentToLocal(handle)->Get(NanSymbol("callback"))
                                             ^~~~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:181:38: note: expanded from macro 'NanSymbol'
#define NanSymbol(value) v8::String::NewSymbol(value)
                         ~~~~~~~~~~~~^
/Users/iliakan/.node-gyp/0.11.13/deps/v8/include/v8.h:1441:8: note: 'IsSymbol' declared here
  bool IsSymbol() const;
       ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make: *** [Debug/obj.target/pathwatcher/src/main.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/iliakan/.nvm/v0.11.13/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:267:23)
gyp ERR! stack     at ChildProcess.EventEmitter.emit (events.js:110:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1046:12)
gyp ERR! System Darwin 13.3.0
gyp ERR! command "node" "/Users/iliakan/.nvm/v0.11.13/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /js/javascript-nodejs/node_modules/gaze
gyp ERR! node -v v0.11.13
gyp ERR! node-gyp -v v0.13.0
gyp ERR! not ok
Build failed
npm ERR! gaze@0.6.4 install: `node build.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the gaze@0.6.4 install script.
npm ERR! This is most likely a problem with the gaze package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node build.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls gaze
npm ERR! There is likely additional logging output above.

npm ERR! System Darwin 13.3.0
npm ERR! command "/Users/iliakan/.nvm/v0.11.13/bin/node" "/Users/iliakan/.nvm/v0.11.13/bin/npm" "i" "git://github.com/shama/gaze.git"
npm ERR! cwd /js/javascript-nodejs
npm ERR! node -v v0.11.13
npm ERR! npm -v 1.4.9
npm ERR! code ELIFECYCLE
npm ERR!
npm ERR! Additional logging details can be found in:
npm ERR!     /js/javascript-nodejs/npm-debug.log
npm ERR! not ok code 0
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,6 +2,7 @@ language: node_js
 node_js:
   - '0.8'
   - '0.10'
+  - '0.11'
 install:
   - npm update npm -g
   - npm install

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -14,10 +14,11 @@ version: "{build}"
 environment:
   matrix:
     - nodejs_version: 0.10
+    - nodejs_version: 0.11
 
 # Get the latest stable version of Node 0.STABLE.latest
 install:
-  - ps: Update-NodeJsInstallation (Get-NodeJsLatestBuild $env:nodejs_version)
+  - ps: Install-Product node $env:nodejs_version
   - npm install
 
 build: off

--- a/package.json
+++ b/package.json
@@ -32,7 +32,7 @@
     "globule": "~0.2.0",
     "nextback": "~0.1.0",
     "bindings": "~1.2.0",
-    "nan": "~0.8.0",
+    "nan": "1.2.0",
     "absolute-path": "0.0.0",
     "graceful-fs": "~2.0.3"
   },

--- a/src/common.cc
+++ b/src/common.cc
@@ -35,35 +35,39 @@ static void CommonThread(void* handle) {
   PlatformThread();
 }
 
+#if NODE_VERSION_AT_LEAST(0, 11, 13)
+static void MakeCallbackInMainThread(uv_async_t* handle) {
+#else
 static void MakeCallbackInMainThread(uv_async_t* handle, int status) {
+#endif
   NanScope();
 
   if (!g_callback.IsEmpty()) {
     Handle<String> type;
     switch (g_type) {
       case EVENT_CHANGE:
-        type = String::New("change");
+        type = NanNew("change");
         break;
       case EVENT_DELETE:
-        type = String::New("delete");
+        type = NanNew("delete");
         break;
       case EVENT_RENAME:
-        type = String::New("rename");
+        type = NanNew("rename");
         break;
       case EVENT_CHILD_CREATE:
-        type = String::New("child-create");
+        type = NanNew("child-create");
         break;
       case EVENT_CHILD_CHANGE:
-        type = String::New("child-change");
+        type = NanNew("child-change");
         break;
       case EVENT_CHILD_DELETE:
-        type = String::New("child-delete");
+        type = NanNew("child-delete");
         break;
       case EVENT_CHILD_RENAME:
-        type = String::New("child-rename");
+        type = NanNew("child-rename");
         break;
       default:
-        type = String::New("unknown");
+        type = NanNew("unknown");
         fprintf(stderr, "Got unknown event: %d\n", g_type);
         return;
     }
@@ -71,11 +75,10 @@ static void MakeCallbackInMainThread(uv_async_t* handle, int status) {
     Handle<Value> argv[] = {
         type,
         WatcherHandleToV8Value(g_handle),
-        String::New(g_new_path.data(), g_new_path.size()),
-        String::New(g_old_path.data(), g_old_path.size()),
+        NanNew(g_new_path.data(), g_new_path.size()),
+        NanNew(g_old_path.data(), g_old_path.size()),
     };
-    NanPersistentToLocal(g_callback)->Call(
-        Context::GetCurrent()->Global(), 4, argv);
+    NanNew(g_callback)->Call(NanGetCurrentContext()->Global(), 4, argv);
   }
 
   WakeupNewThread();
@@ -115,7 +118,7 @@ NAN_METHOD(SetCallback) {
   if (!args[0]->IsFunction())
     return NanThrowTypeError("Function required");
 
-  NanAssignPersistent(Function, g_callback, Handle<Function>::Cast(args[0]));
+  NanAssignPersistent(g_callback, Local<Function>::Cast(args[0]));
   NanReturnUndefined();
 }
 

--- a/src/common.h
+++ b/src/common.h
@@ -37,7 +37,7 @@ bool IsV8ValueWatcherHandle(Handle<Value> value);
 #else
 // Correspoding definetions on OS X and Linux.
 typedef int32_t WatcherHandle;
-#define WatcherHandleToV8Value(h) Integer::New(h)
+#define WatcherHandleToV8Value(h) NanNew<Integer>(h)
 #define V8ValueToWatcherHandle(v) v->Int32Value()
 #define IsV8ValueWatcherHandle(v) v->IsInt32()
 #endif

--- a/src/handle_map.cc
+++ b/src/handle_map.cc
@@ -38,14 +38,14 @@ bool HandleMap::Erase(WatcherHandle key) {
   if (iter == map_.end())
     return false;
 
-  NanDispose(iter->second); // Deprecated, use NanDisposePersistent when v0.12 lands
+  NanDisposeUnsafePersistent(iter->second); // Deprecated, use NanDisposePersistent when v0.12 lands
   map_.erase(iter);
   return true;
 }
 
 void HandleMap::Clear() {
   for (Map::iterator iter = map_.begin(); iter != map_.end(); ++iter)
-    NanDispose(iter->second); // Deprecated, use NanDisposePersistent when v0.12 lands
+    NanDisposeUnsafePersistent(iter->second); // Deprecated, use NanDisposePersistent when v0.12 lands
   map_.clear();
 }
 
@@ -69,7 +69,7 @@ NAN_METHOD(HandleMap::Add) {
   if (obj->Has(key))
     return NanThrowError("Duplicate key");
 
-  NanAssignUnsafePersistent(Value, obj->map_[key], args[1]);
+  NanAssignUnsafePersistent(obj->map_[key], args[1]);
   NanReturnUndefined();
 }
 
@@ -85,7 +85,7 @@ NAN_METHOD(HandleMap::Get) {
   if (!obj->Has(key))
     return NanThrowError("Invalid key");
 
-  NanReturnValue(NanPersistentToLocal(obj->map_[key]));
+  NanReturnValue(NanUnsafePersistentToLocal(obj->map_[key]));
 }
 
 // static
@@ -96,7 +96,7 @@ NAN_METHOD(HandleMap::Has) {
     return NanThrowTypeError("Bad argument");
 
   HandleMap* obj = ObjectWrap::Unwrap<HandleMap>(args.This());
-  NanReturnValue(Boolean::New(obj->Has(V8ValueToWatcherHandle(args[0]))));
+  NanReturnValue(NanNew<Boolean>(obj->Has(V8ValueToWatcherHandle(args[0]))));
 }
 
 // static
@@ -106,11 +106,11 @@ NAN_METHOD(HandleMap::Values) {
   HandleMap* obj = ObjectWrap::Unwrap<HandleMap>(args.This());
 
   int i = 0;
-  Handle<Array> keys = Array::New(obj->map_.size());
+  Handle<Array> keys = NanNew<Array>(obj->map_.size());
   for (Map::const_iterator iter = obj->map_.begin();
        iter != obj->map_.end();
        ++iter, ++i)
-    keys->Set(i, NanPersistentToLocal(iter->second));
+    keys->Set(i, NanUnsafePersistentToLocal(iter->second));
 
   NanReturnValue(keys);
 }
@@ -143,9 +143,9 @@ NAN_METHOD(HandleMap::Clear) {
 void HandleMap::Initialize(Handle<Object> target) {
   NanScope();
 
-  Local<FunctionTemplate> t = FunctionTemplate::New(HandleMap::New);
+  Local<FunctionTemplate> t = NanNew<FunctionTemplate>(HandleMap::New);
   t->InstanceTemplate()->SetInternalFieldCount(1);
-  t->SetClassName(NanSymbol("HandleMap"));
+  t->SetClassName(NanNew("HandleMap"));
 
   NODE_SET_PROTOTYPE_METHOD(t, "add", Add);
   NODE_SET_PROTOTYPE_METHOD(t, "get", Get);
@@ -154,5 +154,5 @@ void HandleMap::Initialize(Handle<Object> target) {
   NODE_SET_PROTOTYPE_METHOD(t, "remove", Remove);
   NODE_SET_PROTOTYPE_METHOD(t, "clear", Clear);
 
-  target->Set(NanSymbol("HandleMap"), t->GetFunction());
+  target->Set(NanNew("HandleMap"), t->GetFunction());
 }

--- a/src/unsafe_persistent.h
+++ b/src/unsafe_persistent.h
@@ -24,59 +24,45 @@ CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #include "nan.h"
 
 #if NODE_VERSION_AT_LEAST(0, 11, 0)
-template<typename TypeName> class NanUnsafePersistent {
- public:
-  NanUnsafePersistent() : value(0) { }
-  explicit NanUnsafePersistent(v8::Persistent<TypeName>* handle) {
-    value = handle->ClearAndLeak();
+template<class T>
+struct NanUnsafePersistentTraits {
+  typedef v8::Persistent<T, NanUnsafePersistentTraits<T> > HandleType;
+  static const bool kResetInDestructor = false;
+  template<class S, class M>
+  static V8_INLINE void Copy(const Persistent<S, M>& source, 
+                             HandleType* dest) {
+    // do nothing, just allow copy
   }
-  explicit NanUnsafePersistent(const v8::Local<TypeName>& handle) {
-    v8::Persistent<TypeName> persistent(nan_isolate, handle);
-    value = persistent.ClearAndLeak();
-  }
-
-  NAN_INLINE(v8::Persistent<TypeName>* persistent()) {
-    v8::Persistent<TypeName>* handle = reinterpret_cast<v8::Persistent<TypeName>*>(&value);
-    return handle;
-  }
-
-  NAN_INLINE(void Dispose()) {
-    NanDispose(*persistent());
-    value = 0;
-  }
-
-  NAN_INLINE(void Clear()) {
-    value = 0;
-  }
-
-  NAN_INLINE(v8::Local<TypeName> NewLocal()) {
-    return v8::Local<TypeName>::New(nan_isolate, *persistent());
-  }
-
-  NAN_INLINE(bool IsEmpty() const) {
-    return value;
-  }
-
- private:
-  TypeName* value;
 };
-#define NanAssignUnsafePersistent(type, handle, obj)                          \
-  handle = NanUnsafePersistent<type>(obj)
-template<class T> static NAN_INLINE(void NanDispose(
-    NanUnsafePersistent<T> &handle
-)) {
-  handle.Dispose();
-  handle.Clear();
+
+template<class T>
+class NanUnsafePersistent : public NanUnsafePersistentTraits<T>::HandleType {
+  public:
+    V8_INLINE NanUnsafePersistent() {}
+    template <class S>
+      V8_INLINE NanUnsafePersistent(v8::Isolate* isolate, S that)
+        : NanUnsafePersistentTraits<T>::HandleType(isolate, that) {}
+};
+
+template<typename T>
+  NAN_INLINE void NanAssignUnsafePersistent(
+    NanUnsafePersistent<T>& handle
+    , v8::Handle<T> obj) {
+    handle.Reset();
+    handle = NanUnsafePersistent<T>(v8::Isolate::GetCurrent(), obj);
 }
-template <class TypeName>
-static NAN_INLINE(v8::Local<TypeName> NanPersistentToLocal(
-   const NanUnsafePersistent<TypeName>& persistent
-)) {
-  return const_cast<NanUnsafePersistent<TypeName>&>(persistent).NewLocal();
-}
+
+template<typename T>
+  NAN_INLINE v8::Local<T> NanUnsafePersistentToLocal(const NanUnsafePersistent<T> &arg1) {
+    return v8::Local<T>::New(v8::Isolate::GetCurrent(), arg1);
+  }
+
+#define NanDisposeUnsafePersistent(handle) handle.Reset()
 #else
 #define NanUnsafePersistent v8::Persistent
 #define NanAssignUnsafePersistent NanAssignPersistent
+#define NanUnsafePersistentToLocal NanNew
+#define NanDisposeUnsafePersistent NanDisposePersistent
 #endif
 
 #endif  // UNSAFE_PERSISTENT_H_

